src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
> I wouldn't mind moving it in the OWL API.

Splendid.  For 4.1?!? :)

>  already does the required locking, the convenience/legacy variants can delegate without loss.

I think it's clearer if everything tries to acquire a read lock.  The locks are re-entrant anyway.  Perhaps there is a performance problem with this, but I don't really know.

> If you wish to move it to OWL API, I don't have objections - we can add it as a new module. Could be this code, or equivalent code if just moving this code could cause copyright owner headaches in Stanford.

I will move it over.  No copyright problems.